metronome/rocksdb/src/io/iohk/metronome/rocksdb/RocksDBStore.scala

```diff
@@ -0,0 +1,374 @@
+package io.iohk.metronome.rocksdb
+
+import cats._
+import cats.implicits._
+import cats.data.ReaderT
+import cats.effect.{Resource, Sync}
+import cats.free.Free.liftF
+import io.iohk.metronome.storage.{KVStore, KVStoreOp}
+import io.iohk.metronome.storage.KVStoreOp.{Put, Get, Delete}
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import org.rocksdb.{
+  RocksDB,
+  WriteBatch,
+  WriteOptions,
+  ReadOptions,
+  Options,
+  DBOptions,
+  ColumnFamilyOptions,
+  ColumnFamilyDescriptor,
+  ColumnFamilyHandle,
+  BlockBasedTableConfig,
+  BloomFilter,
+  CompressionType,
+  ClockCache
+}
+import scodec.Codec
+import scodec.bits.BitVector
+import scala.collection.mutable
+import java.nio.file.Path
+import scala.annotation.nowarn
+
+class RocksDBStore[F[_]: Sync](
+    db: RocksDB,
+    readOptions: ReadOptions,
+    rwlock: ReentrantReadWriteLock,
```

Comment:
Thanks for the explantation!

Although I still think that this is overly restrictive from such low level layer.  I feel like it is optimising for problems we had in another project with different data model, at the expense of ease of usage and performance of basic operations.

Consider:
- now even point reads and point writes need to go through this lock, which is not necessary.
- writing unrelated data to different columns in parallel is not possible, as there is one write lock.

I think that at this level it would be better to expose api which make it possible to create write batch which would return something along `Resource[F, WriteBatch]` and build all this batching and synchronisation in upper layer using higher level constructs.

It maybe only my opinion, so I will not press harder for it if you are still not convinced.

Reply:
These are great points! Note, however, that this is just one of the ways a `KVStore[N, A]` can be interpreted, and turning the locking on/off is as easy as calling `runWithBatchingNoLock`. This is not the interface that the user code has to deal with directly, you can pass around any interpreter function to the code that you want, and compose it in your application root (only `app` has a dependency on `rocksdb` in `build.sc`), so if you, as a develop, know that your data store is append-only, therefore reads don't have to be isolated from writes, and also that writes can be done in parallel, then by all means use the interpreters that don't use locks. Indeed with HotStuff we should never have rollbacks, so the problem where the data is gone may never arise.

You are right that this design is inspired by the problem we had in other projects, where in a PR that made a futile attempt at adding transactions I suggested that turning the patterns upside down or using the Free monad would help, and I wanted to see how that would work. The good thing is that the _what_ is completely separate from the _how_ in this case, so treat is as a demonstration of what's possible. 

I don't really like the names of the methods `runWithBatching` and `runWithoutBatching`, but I have difficulty coming up with better ones that capture their semantics. To hopefully put you more at ease, I'll add more methods that don't use locks, so it's clearer from the API that all these are options.

Reply:
Done, now you can call the methods without locking, and I added comments to explain that you're expected to roll your own adapter layer that picks the right strategy, or does its own locking.